mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -2389,6 +2650,87 @@ struct CopyToDmaPass : public air::impl::CopyToDmaBase<CopyToDmaPass> {
 
     LLVM_DEBUG(llvm::outs() << "output\n");
     LLVM_DEBUG(module.print(llvm::outs()));
+
+    // Condense memref data pattern reordering ops, including memref.subview,
+    // memref.tranpose and memref.expand_shape into air.dma_memcpy_nd op's
+    // offsets, sizes and strides fields.
+    auto scope = getOperation();
+    std::vector<std::tuple<xilinx::air::DmaMemcpyNdOp, std::vector<Operation *>,
+                           std::vector<Operation *>>>
+        dma_ops;
+
+    scope->walk([&](xilinx::air::DmaMemcpyNdOp dmaOp) {
+      bool src_condense = false;
+      if (auto src_defop = dmaOp.getSrcMemref().getDefiningOp()) {
+        src_condense |= isa<memref::TransposeOp>(src_defop);
+        src_condense |= isa<memref::ExpandShapeOp>(src_defop);
+        src_condense |= isa<memref::SubViewOp>(src_defop);
+      }
+      bool dst_condense = false;
+      if (auto dst_defop = dmaOp.getDstMemref().getDefiningOp()) {
+        dst_condense |= isa<memref::TransposeOp>(dst_defop);
+        dst_condense |= isa<memref::ExpandShapeOp>(dst_defop);
+        dst_condense |= isa<memref::SubViewOp>(dst_defop);
+      }
+      if (src_condense || dst_condense) {
+        std::tuple<air::DmaMemcpyNdOp, std::vector<Operation *>,
+                   std::vector<Operation *>>
+            log_entry;
+        std::get<0>(log_entry) = dmaOp;
+        if (src_condense) {
```

Comment:
nit: I think the number of lines can be reduced by not duplicating the code for src and dst cases (and less code is generally good :))